### PR TITLE
[optim] Add __repr__ to LRScheduler so schedulers print their settings

### LINKED ISSUE #168967: implement __str__ for Schedular classes
### 🚀 The feature, motivation and pitch

I'm logging my setup to a txt file, including model/optimizer/scheduler, and would like the string representation of schedulers to be meaningful, similar to `nn.Module` and Pytorch's Optimizers.

### Alternatives

_No response_

### Additional context

logging a scheduler: 
```
<torch.optim.lr_scheduler.OneCycleLR object at 0x14b688123e30>
```
, which is not meaningful, while logging the optimizers
```
AdamW (
Parameter Group 0
    amsgrad: False
    base_momentum: 0.85
    betas: (0.95, 0.999)
    capturable: False
    decoupled_weight_decay: True
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: None
    initial_lr: 4e-05
    lr: 3.9999999999999996e-05
    max_lr: 0.001
    max_momentum: 0.95
    maximize: False
    min_lr: 4e-09
    weight_decay: 0.01
)
```

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by eyad6789**
Fixes #168967

## Summary

Printing an LR scheduler gives you the default object repr:

```
<torch.optim.lr_scheduler.OneCycleLR object at 0x14b688123e30>
```

Optimizers already print their param groups, so a setup log ends up showing the optimizer in full and the scheduler as an address — which is exactly what the issue reports.

This adds `LRScheduler.__repr__`, following the `nn.Module` pattern: class name, then the settings named by a new `_extra_repr()` hook, then `last_epoch` and `base_lrs`. Each scheduler overrides `_extra_repr()` to name the parameters that define its schedule.

After:

```
OneCycleLR (                       StepLR (                      ReduceLROnPlateau (
    total_steps: 100                   step_size: 30                 mode: min
    cycle_momentum: False              gamma: 0.1                    factor: 0.5
    last_epoch: 0                      last_epoch: 0                 patience: 5
    base_lrs: [4e-05]                  base_lrs: [0.05]              threshold: 0.0001
)                                  )                                 threshold_mode: rel
                                                                     cooldown: 0
                                                                     min_lrs: [0]
                                                                     eps: 1e-08
                                                                     last_epoch: 0
                                                                 )
```

A scheduler that defines no `_extra_repr()` — `SWALR`, or any user subclass — still gets the base representation.

## Note on the two optional fields

`last_epoch` and `base_lrs` are emitted only when the instance actually has them. `ChainedScheduler` sets neither, and `SequentialLR` and `ReduceLROnPlateau` never set `base_lrs`, because none of the three route through `LRScheduler.__init__`. Guarding on the attributes is what keeps `repr()` from raising on those three rather than defensive habit — `test_lr_scheduler_repr_nests_composite_schedulers` pins that behaviour.

## Testing

Every public scheduler is covered by one table-driven case rather than one test per class, plus focused cases for the tensor-valued `base_lrs` path, the composite schedulers that omit those fields, the momentum branch of `CyclicLR`, and a subclass that defines no `_extra_repr()`.

Verified against `torch==2.15.0.dev20260821+cpu` with the modified `torch/optim/lr_scheduler.py` overlaid into the wheel (this machine does not have a source build of the C++ side):

- new tests, unmodified wheel: `20 failed, 1 passed` — every new assertion fails without the change
- new tests, modified wheel: `5 passed, 16 subtests passed`
- whole file, modified wheel: `python -m pytest test/optim/test_lrscheduler.py` → `196 passed, 1 xfailed, 16 subtests passed`

Lint is clean under `ruff==0.14.4`, the version pinned in `tools/linter/adapters/pyfmt_linter.py`: `ruff format --diff` reports both files already formatted and `ruff check` passes.

## Relationship to prior work

This relands #169085 by @tshauck, which you approved and tried to merge three times before it failed lint and went stale. The credit for the original implementation is theirs. Two things are changed on top of it, both from your review comments there and on the parallel attempt in #176736:

- `extra_repr` is renamed to `_extra_repr`, as you asked
- `SequentialLR` and `ChainedScheduler` report their nested schedulers, and the base `__repr__` tolerates their missing attributes

The lint failure that stalled the original is fixed and verified locally with the pinned ruff.

cc @janeyx99 @albanD @crcrpar

---

Prepared with AI assistance. I have read and reviewed the analysis and the implementation, and I stand behind both.

### CODE REVIEW

**Comment by janeyx99** [comment]
@claude please review this and verify that all docs are accurate to the Scheduler

**Comment by janeyx99** [comment]
@eyad6789 ah it seems Claude is catching some good points here. Would you be willing to address them?

**Comment by eyad6789** [comment]
@janeyx99 Yes, happy to. All four hold up when I check them against the code — I'll push a follow-up commit

**Comment by eyad6789** [comment]
@janeyx99 Apologies for the silence -- I pushed the follow-up in 78473f11f right after my last comment and never came back to say so. All four points are addressed:

1. **Nested composites** rendered through `list.__repr__`, which spliced `[`, `]` and `, ` into the middle of each child's multi-line block and indented every depth the same. `SequentialLR` and `ChainedScheduler` now build the block via `_schedulers_repr`, which indents each child once so the base `__repr__`'s own indent compounds correctly at arbitrary depth.
2. **`LambdaLR` / `MultiplicativeLR`** printed `<function <lambda> at 0x...>`, which is the address the linked issue is about. They now print each callable's `__qualname__`, falling back to `type(fn).__qualname__` for a callable object, which has none of its own.
3. **`OneCycleLR`** never showed `max_lr` -- it lives on the param groups, and `base_lrs` holds `max_lr / div_factor`. It now reports `max_lrs`, `pct_start`, `anneal_strategy`, `three_phase` and the momentum bounds. The param-group reads are guarded because `OneCycleLR` only writes those keys when `last_epoch == -1`.
4. **`CyclicLR`** printed `mode` and `gamma` even when a custom `scale_fn` made both inert -- `mode` isn't validated in that case and `gamma` is never read -- so it reports `scale_fn` instead.

Also added `SWALR._extra_repr`, and widened the test's address guard from `object at 0x` to `at 0x`. A function repr never contains "object", so the old guard was missing exactly the two schedulers that leaked an address.

On the question I left open last time: I've kept `_extra_repr` to constructor settings only, so it does not report `T_i`/`T_cur` for `CosineAnnealingWarmRestarts` or `default_min_lr` for `ReduceLROnPlateau`. That keeps `repr()` stable across `step()` and matches how `nn.Module.extra_repr` behaves. Easy to change if you'd rather see live state.

One thing I can't do myself: CI has never run on this PR. Both commits show zero workflow runs and EasyCLA is the only status, so the workflows are still awaiting approval. Could you trigger them when you get a chance?

Separately, while working in this file I noticed #128046 is still open with no PR against it, and its no-inplace goal comes down to the single `fill_` left in `_update_param_group_val`. I'd like to take it, but replacing that with assignment changes behaviour for anyone holding a reference to the LR tensor, which is the whole reason the helper exists. Is that the direction you want, or is the aliasing intentional? Happy to follow up with a PR if so.

**Comment by eyad6789** [comment]
@janeyx99 Apologies for the silence — I pushed the follow-up in 78473f11f right
after my last comment and never came back to say so. All four points are addressed:

1. **Nested composites** rendered through `list.__repr__`, which spliced `[`, `]`
   and `, ` into the middle of each child's multi-line block and indented every
   depth the same. `SequentialLR` and `ChainedScheduler` now build the block via
   `_schedulers_repr`, which indents each child once so the base `__repr__`'s own
   indent compounds correctly at arbitrary depth.
2. **`LambdaLR` / `MultiplicativeLR`** printed `<function <lambda> at 0x...>`,
   which is the address the linked issue is about. They now print each callable's
   `__qualname__`, falling back to `type(fn).__qualname__` for a callable object,
   which has none of its own.
3. **`OneCycleLR`** never showed `max_lr` — it lives on the param groups, and
   `base_lrs` holds `max_lr / div_factor`. It now reports `max_lrs`, `pct_start`,
   `anneal_strategy`, `three_phase` and the momentum bounds. The param-group reads
   are guarded because `OneCycleLR` only writes those keys when `last_epoch == -1`.
4. **`CyclicLR`** printed `mode` and `gamma` even when a custom `scale_fn` made
   both inert — `mode` isn't validated in that case and `gamma` is never read — so
   it reports `scale_fn` instead.

Also added `SWALR._extra_repr`, and widened the test's address guard from
`object at 0x` to `at 0x`. A function repr never contains "object", so the old
guard was missing exactly the two schedulers that leaked an address.

On the question I left open last time: I've kept `_extra_repr` to constructor
settings only, so it does not report `T_i`/`T_cur` for
`CosineAnnealingWarmRestarts` or `default_min_lr` for `ReduceLROnPlateau`. That
keeps `repr()` stable across `step()` and matches how `nn.Module.extra_repr`
behaves. Easy to change if you'd rather see live state.

One thing I can't do myself: **CI has never run on this PR.** Both commits show
zero workflow runs and EasyCLA is the only status, so the workflows are still
awaiting approval. Could you trigger them when you get a chance?

Separately, while working in this file I noticed #128046 is still open with no PR
against it, and its no-inplace goal comes down to the single `fill_` left in
`_update_param_group_val`. I'd like to take it, but replacing that with assignment
changes behaviour for anyone holding a reference to the LR tensor, which is the
whole reason the helper exists. Is that the direction you want, or is the aliasing
intentional? Happy to follow up with a PR if so.

**Comment by eyad6789** [comment]
@janeyx99 there's a concrete blocker on this one that isn't the review: **CI has never run.**

All four workflows (Lint, BC Lint, docs-build, pull) have been sitting in `action_required`
since 2026-08-25 waiting for maintainer approval — the only two green checks on the PR are
EasyCLA and the Meta internal check. Could you hit "Approve and run workflows"? It can't
merge from here regardless of how the review goes.

The four points from the `@claude` review are addressed in `78473f11f` (pushed 08-25;
itemised in my comment above). Happy to rebase or split it if that's easier to review.

One thing still open from that thread and it's your call: should `_extra_repr` stay
settings-only, or may it report live state — `T_i`/`T_cur` for `CosineAnnealingWarmRestarts`,
`default_min_lr` for `ReduceLROnPlateau`? I kept it settings-only.
